src/deserialize/parse.rs

```diff
@@ -0,0 +1,361 @@
+            Ok(Edn::Bool(string.parse::<bool>()?))
+        }
+        'n' => {
+            let c_len = chars
+                .clone()
+                .take_while(|e| e == &'i' || e == &'l')
+                .collect::<String>()
+                .len();
+            let mut string = String::new();
+            let n = chars.take(c_len).collect::<String>();
+            string.push(c);
+            string.push_str(&n);
+            match &string[..] {
+                "nil" => Ok(Edn::Nil),
+                _ => Err(Error::ParseEdn(format!("{} cound not be parsed", string))),
+            }
+        }
+        _ => Err(Error::ParseEdn(
+            "Nullable boolean cound not be parsed".to_string(),
+        )),
+    }
+}
+
+fn read_vec(chars: &mut std::str::Chars) -> Result<Edn, Error> {
+    let mut res: Vec<Edn> = vec![];
+    loop {
+        match chars.next() {
+            Some(']') => return Ok(Edn::Vector(Vector::new(res))),
+            Some(c) if !c.is_whitespace() && c != ',' => {
+                res.push(parse(Some(c), chars)?);
+            }
+            Some(c) if c.is_whitespace() || c == ',' => (),
+            err => return Err(Error::ParseEdn(format!("{:?} could not be parsed", err))),
+        }
+    }
+}
+
+fn read_list(chars: &mut std::str::Chars) -> Result<Edn, Error> {
+    let mut res: Vec<Edn> = vec![];
+    loop {
+        match chars.next() {
+            Some(')') => return Ok(Edn::List(List::new(res))),
+            Some(c) if !c.is_whitespace() && c != ',' => {
+                res.push(parse(Some(c), chars)?);
+            }
+            Some(c) if c.is_whitespace() || c == ',' => (),
+            err => return Err(Error::ParseEdn(format!("{:?} could not be parsed", err))),
+        }
+    }
+}
+
+fn read_set(chars: &mut std::str::Chars) -> Result<Edn, Error> {
+    use std::collections::BTreeSet;
+    let mut res: BTreeSet<Edn> = BTreeSet::new();
+    loop {
+        match chars.next() {
+            Some('}') => return Ok(Edn::Set(Set::new(res))),
+            Some(c) if !c.is_whitespace() && c != ',' => {
+                res.insert(parse(Some(c), chars)?);
+            }
+            Some(c) if c.is_whitespace() || c == ',' => (),
+            err => return Err(Error::ParseEdn(format!("{:?} could not be parsed", err))),
+        }
+    }
+}
+
+fn read_map(chars: &mut std::str::Chars) -> Result<Edn, Error> {
+    use std::collections::BTreeMap;
+    let mut res: BTreeMap<String, Edn> = BTreeMap::new();
+    let mut key: Option<Edn> = None;
+    let mut val: Option<Edn> = None;
+    loop {
+        match chars.next() {
+            Some('}') => return Ok(Edn::Map(Map::new(res))),
+            Some(c) if !c.is_whitespace() && c != ',' => {
+                if key.is_some() {
+                    val = Some(parse(Some(c), chars)?);
+                } else {
+                    key = Some(parse(Some(c), chars)?);
+                }
+            }
+            Some(c) if c.is_whitespace() || c == ',' => (),
```

Comment:
Maybe removal of unnecessary arm? :thinking:

Reply:
test result: FAILED. 36 passed; 3 failed; 0 ignored; 0 measured; 0 filtered out